### PR TITLE
Add MLE lecture

### LINKED ISSUE #56: Add a lecture on max likelihood estimation
We could redo this discussion in JAX: https://python.quantecon.org/mle.html#mle-with-numerical-methods

We only need to do section 80.5.

First we can use automatic differentiation.  Then we can repeat with the exact derivatives, the same as the existing code, and check that we get the same results.

We can explain that automatic differentiation is easier, more efficient, and possible when the gradients can't be computed analytically.

@Smit-create , do you have time to look at this?

### PR DESCRIPTION
**Pull request by Smit-create**
Fixes #56

### CODE REVIEW

**Comment by Smit-create** [comment]
This is ready for a review @jstac. Thanks!

**Comment by jstac** [comment]
Many thanks @Smit-create , nice work!

You get different numbers for `beta_hat` in this lecture --- different to the values in the original lecture.

Do you know why?

Also, in the code in https://aosmith.rbind.io/2018/07/18/simulate-poisson-edition/, the author simulates from a Poisson model.  If you plug the generated outcomes into your estimator, then, provided that the sample size is large, you should recover the values of $\beta$ (in his case, `b0, b1, b2`) used in the simulation (approximately).

This could be an exercise.  

Would you mind to add that?

File: lectures/mle.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+```
+
+
+The following function computes the `factorial`.
+
+```{code-cell} ipython3
+@jax.jit
+def _factorial(n):
+    return jax.lax.exp(jax.lax.lgamma(n + 1.0)).astype(int)
+
```
Comment (jstac) [inline lectures/mle.md:148]: could you please provide a reference for this so the reader can understand what is happening?

File: lectures/mle.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+To find the gradient of the `poisson_logL`, we again use [jax.grad](https://jax.readthedocs.io/en/latest/_autosummary/jax.grad.html).
+
+To find the Hessian, we can directly use [jax.jacfwd](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jacobians-and-hessians-using-jacfwd-and-jacrev).
+
+```{code-cell} ipython3
+G_poisson_logL = jax.grad(poisson_logL)
+H_poisson_logL = jax.jacfwd(G_poisson_logL)
+```
```
Comment (jstac) [inline lectures/mle.md:171]: Perhaps add one line saying that the Hessian is the Jacobian of the gradient?

Why do we use `jacfwd` in the second derivative?  Why note `jax.jacobian`?

**Comment by Smit-create** [comment]
> Do you know why?

I'm not exactly sure, we can try using double precision in JAX and compare again.

File: lectures/mle.md (hunk 2 above)
Comment (Smit-create) [reply lectures/mle.md:171]: > Why do we use jacfwd in the second derivative?

I referred that using this section of the docs: https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jacobians-and-hessians-using-jacfwd-and-jacrev. 

Moreover, from this issue: https://github.com/google/jax/issues/2344, it says: "Note that `jacobian` is currently just an alias for `jaxrev`."
.

Comment (jstac) [reply lectures/mle.md:171]: Thanks @Smit-create .  Actually this discussion is very helpful and I suggest we add it and then connect it to our choice to use `jaxfwd` in this particular case.  We can write:

According to [the documentation](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jacobians-and-hessians-using-jacfwd-and-jacrev),

* `jacfwd` uses forward-mode automatic differentiation, which is more efficient for “tall” Jacobian matrices, while 
* `jacrev` uses reverse-mode, which is more efficient for “wide” Jacobian matrices. 
 
(The documentation also states that when matrices that are near-square, `jacfwd` probably has an edge over `jacrev`.)

Comment (Smit-create) [reply lectures/mle.md:198]: Thanks, added that.

**Comment by Smit-create** [comment]
> You get different numbers for beta_hat in this lecture --- different to the values in the original lecture.
Do you know why?

Fixed this issue. There was a small reshaping bug. Thanks for pointing.

File: lectures/mle.md (hunk 1 above)
Comment (Smit-create) [reply lectures/mle.md:168]: Added a small explanation above this cell. Please see if that's okay.

**Comment by jstac** [comment]
Nice work @Smit-create !

Can you please fix the rendering in 

image

and use `\log` for the log function

**Comment by jstac** [comment]
Thanks @Smit-create , merging.

**Comment by Smit-create** [comment]
Many thanks @jstac. Should we add this as an exercise?

"""
Also, in the code in https://aosmith.rbind.io/2018/07/18/simulate-poisson-edition/, the author simulates from a Poisson model. If you plug the generated outcomes into your estimator, then, provided that the sample size is large, you should recover the values of 
 (in his case, b0, b1, b2) used in the simulation (approximately).

This could be an exercise.
"""

I can do that in a new PR.

**Comment by jstac** [comment]
Thanks @Smit-create , I forgot about that.

It would be great if you could address that in a new PR.
